packages/bsky/src/api/app/bsky/notification/listNotifications.ts

```diff
@@ -43,6 +43,56 @@ export default function (server: Server, ctx: AppContext) {
   })
 }
 
+const paginateNotifications = async (opts: {
+  ctx: Context
+  priority: boolean
+  filter?: string[]
+  cursor?: string
+  limit: number
+  viewer: string
+}) => {
+  const { ctx, priority, filter, limit, viewer } = opts
+
+  // if not filtering, then just pass through the response from dataplane
+  if (!filter) {
+    const res = await ctx.hydrator.dataplane.getNotifications({
+      actorDid: viewer,
+      priority,
+      cursor: opts.cursor,
+      limit,
+    })
+    return {
+      notifications: res.notifications,
+      cursor: res.cursor,
+    }
+  }
+
+  let nextCursor: string | undefined = opts.cursor
+  let toReturn: Notification[] = []
+  const maxAttempts = 10
+  const attemptSize = Math.ceil(limit / 2)
+  for (let i = 0; i < maxAttempts; i++) {
+    const res = await ctx.hydrator.dataplane.getNotifications({
+      actorDid: viewer,
+      priority,
+      cursor: nextCursor,
+      limit,
```

Comment:
Yeah I think we can improve this in the future by passing the filter to the dataplane, but this is the much simpler impl for now.

I'm not overly concerned about he performance on this and I expect filtered queries to be quite a bit lower throughput than `listNotifications` in general. The reason to not request extra is that we may end up with more posts than the requested `limit` and should truncate the result set as a result which would change the cursor that we're returning. Currently we pass through the cursor right from the dataplane and the appview codebase doesn't have a semantic understanding of the cursor so it's not able to create a new cursor based on the truncated result set. Think there's a good chance that in the future we make cursors better specified so that the appview can correctly format them itself